File: frontend/src/api/aiFeedback.ts
import { apiClient } from './client';

export interface AIFeedbackCreate {
  entity_type: string;
  entity_id?: string;
  field_name: string;
  original_value: string;
  corrected_value: string;
  model_version?: string;
  notes?: string;
}

export interface AIFeedbackRead {
  id: string;
  feedback_type: string;
  entity_type: string;
  entity_id: string;
  field_name: string | null;
  original_value: string | null;
  corrected_value: string | null;
  confidence_delta: number | null;
  model_version: string | null;
  user_id: string;
  notes: string | null;
  created_at: string;
}

export interface CommonErrorEntry {
  original: string;
  corrected: string;
  count: number;
}

export interface AIMetricsRead {
  id: string;
  entity_type: string;
  field_name: string;
  total_extractions: number;
  total_corrections: number;
  error_rate: number;
  common_errors: CommonErrorEntry[];
  updated_at: string;
}

export interface AIMetricsSummary {
  overall_accuracy: number;
  total_extractions: number;
  total_corrections: number;
  metrics: AIMetricsRead[];
}

export const aiFeedbackApi = {
  recordFeedback: (diagnosticId: string, data: AIFeedbackCreate) =>
    apiClient.post<AIFeedbackRead>(`/diagnostics/${diagnosticId}/feedback`, data),

  getMetrics: (entityType?: string) =>
    apiClient.get<AIMetricsSummary>('/analytics/ai-metrics', {
      params: entityType ? { entity_type: entityType } : undefined,
    }),

  getMetricsByType: (entityType: string) =>
    apiClient.get<AIMetricsRead[]>(`/analytics/ai-metrics/${entityType}`),

  listFeedback: (params?: { entity_type?: string; entity_id?: string; limit?: number }) =>
    apiClient.get<AIFeedbackRead[]>('/analytics/ai-feedback', { params }),
};

File: backend/app/schemas/ai_feedback.py
"""Schemas for AI feedback loop — correction recording + metrics."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class AIFeedbackCreate(BaseModel):
    """Record a field-level AI correction."""

    entity_type: str  # diagnostic | material | sample
    entity_id: UUID | None = None  # defaults to diagnostic_id from path
    field_name: str  # e.g. "material_type", "hazard_level"
    original_value: str
    corrected_value: str
    model_version: str | None = None
    notes: str | None = None

    model_config = ConfigDict(from_attributes=True)


class AIFeedbackRead(BaseModel):
    """AI feedback record."""

    id: UUID
    feedback_type: str
    entity_type: str
    entity_id: UUID
    field_name: str | None = None
    original_value: str | None = None
    corrected_value: str | None = None
    confidence_delta: float | None = None
    model_version: str | None = None
    user_id: UUID
    notes: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CommonErrorEntry(BaseModel):
    original: str
    corrected: str
    count: int


class AIMetricsRead(BaseModel):
    """Aggregated AI accuracy metrics per entity_type + field_name."""

    id: UUID
    entity_type: str
    field_name: str
    total_extractions: int
    total_corrections: int
    error_rate: float
    common_errors: list[CommonErrorEntry] = []
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AIMetricsSummary(BaseModel):
    """Dashboard-level summary across all fields."""

    overall_accuracy: float
    total_extractions: int
    total_corrections: int
    metrics: list[AIMetricsRead] = []

    model_config = ConfigDict(from_attributes=True)
